ehsm_kms_service/secret_manager_apis.js

```diff
@@ -265,9 +265,104 @@ const createSecret = async (res, appid, payload, DB) => {
     }
 }
 
+//stores the secret value of a new version into a secret object.
+const putSecretValue = async (res, appid, payload, DB) => {
+    try {
+        //get and check Param in payload
+        let secretName = getParam_String(payload, 'secretName')
+        let secretData = getParam_String(payload, 'secretData')
+        const createTime = new Date().getTime()
+        let sm_masterKey
+        let encryptionKeyId
+        let versionId
+        if (!checkStringParam(secretData, true, SECRETDATA_LENGTH_MAX)) {
+            res.send(_result(400, `secretData cannot be empty, must be string and length not more than ${SECRETDATA_LENGTH_MAX}`))
+            return
+        }
+        if (!checkStringParam(secretName, true, SECRETNAME_LENGTH_MAX)) {
+            res.send(_result(400, `secretName cannot be empty, must be string and length not more than ${SECRETNAME_LENGTH_MAX}`))
+            return
+        }
+
+        //query SM_Masterkey and encrypt secretData
+        const query_encryptionKeyId = {
+            selector: {
+                appid,
+                secretName,
+            },
+            fields: ['encryptionKeyId'],
+            limit: 1,
+        }
+        const secret_metadata_res = await DB.partitionedFind('secret_metadata', query_encryptionKeyId)
+        if (secret_metadata_res.docs.length == 0) {
+            encryptionKeyId = ""
+        } else {
+            encryptionKeyId = secret_metadata_res.docs[0].encryptionKeyId
+        }
+        retCMK = await get_SM_Masterkey(DB, appid, encryptionKeyId)
+        if (retCMK.isFind) {
+            sm_masterKey = retCMK.sm_masterKey
+        } else {
+            res.send(_result(400, retCMK.msg))
+            return
+        }
+        const apikey_encrypt_res = napi_result(cryptographic_apis.Encrypt, res, [sm_masterKey, secretData, ''])
+        const { ciphertext } = apikey_encrypt_res.result
+
+        //Search the old version by appid, secretName, versionStage, deletedFlag 
+        //change old version stage and add a new version
+        const query_version_stage = {
+            selector: {
+                appid,
+                secretName,
+                versionStage: SM_SECRET_VERSION_STAGE_CURRENT,
+                deletedFlag: false
+            },
+            fields: [
+                '_id',
+                '_rev',
+                'appid',
+                'secretName',
+                'versionId',
+                'deletedFlag',
+                'secretData',
+                'createTime',
+                'versionStage',
+            ],
+            limit: 1
+        }
+        const version_stage_res = await DB.partitionedFind('secret_version_data', query_version_stage)
+        if (version_stage_res.docs.length > 0) {
+            versionId = version_stage_res.docs[0].versionId + 1
+            version_stage_res.docs[0].versionStage = SM_SECRET_VERSION_STAGE_PREVIOUS
+            await DB.insert(version_stage_res.docs[0])
```

Comment:
use "insert" to update a DB? Can we use “update DB“ code here？

Reply:
no direct update function in CouchDB, can only be replaced by insert

Reply:
if _id is matched, DB.insert will update the same record?  is my understanding correct?

Reply:
Line 353 is to insert a new record OR to update the original one?